python/pysrc/kaskada/udf.py

```diff
@@ -1,57 +1,54 @@
 """Functionality for calling Python UDFs from Kaskada."""
+from __future__ import annotations
+
 import functools
 from typing import Callable
 
 import pandas as pd
 import pyarrow as pa
 
+from ._timestream import Timestream
+from ._timestream import Literal
+from . import _ffi
 
 # TODO: Allow functions to return `pd.DataFrame` for struct arrays.
 FuncType = Callable[..., pd.Series]
 
-
-class Udf(object):
-    """Class wrapping a UDF used in Kaskada."""
-
-    def __init__(self, name, func: FuncType, signature: str) -> None:
-        """Create a UDF for a function returning a Pandas series.
-
-        Parameters
-        ----------
-        name : str
-            Name of the function being wrapped.
-
-        func : FuncType
-            The callable to wrap.
-
-        signature : str
-            The Kaskada function signature for this UDF. Will be used to check
-            types of calls to this function and propagate type information for
-            the rest of the query.
-        """
+class Udf:
+    def __init__(self, signature: str, func: Callable[..., pa.Array]) -> None:
+        """Create a UDF with the given signature."""
         functools.update_wrapper(self, func)
-        self.name = name
-        self.func = func
-        self.signature = signature
-
-    def run_pyarrow(self, result_type: pa.DataType, *args: pa.Array) -> pa.Array:
-        """Run the function producing the given result type."""
-        # TODO: I believe this will return a series for simple arrays, and a
-        # dataframe for struct arrays. We should explore how this handles
-        # different types.
-        pd_args = [arg.to_pandas() for arg in args]
-        pd_result = self.func(*pd_args)
-
-        if isinstance(pd_result, pd.Series):
-            return pa.Array.from_pandas(pd_result, type=result_type)
-        else:
-            raise TypeError(f"Unsupported result type: {type(pd_result)}")
+        self._ffi = _ffi.Udf(signature, func)
 
+    def __call__(self, *args: Timestream | Literal) -> Timestream:
+        """Apply the UDF to the given arguments."""
+        print("Calling udf")
+        return Timestream._call_udf(self._ffi, *args)
```

Comment:
I think this can just use `Timestream._call(self, *args)`, then in timestream you can change the `name: str` parameter to be `func: str | Udf`, and then do a `if isinstance(Udf, func)`